File: python/practice_problems/problem_01_geofence_alert_engine.py
"""
=============================================================================
INTERVIEW PROBLEM 1: Geofence Alert Rule Engine
Difficulty: Senior Software Engineer | Estimated time: 45 min
=============================================================================

CONTEXT
-------
You're building a backend service for an IoT asset-tracking platform. Physical
assets (forklifts, shipping containers, field equipment) carry GPS sensors that
periodically report coordinates. The platform tracks which geographic "zone"
(geofence) each asset is currently inside, and fires configured alert rules
whenever an asset transitions between zones.

For this problem, zones are axis-aligned bounding boxes — no geospatial
libraries needed.

DATA MODEL
----------
All state lives in a single TrackerState dict (returned by make_tracker).

Zone:
  {"id": str, "name": str, "bounds": {"min_lat": float, "max_lat": float,
                                       "min_lng": float, "max_lng": float}}

Asset:
  {"id": str, "name": str, "lat": float | None, "lng": float | None, "zone_id": str | None}
  - lat/lng are None until the first GPS update arrives.
  - zone_id is the id of the zone the asset is currently in, or None.

AlertRule:
  {"id": str, "from_zone_id": str | None, "to_zone_id": str | None, "asset_id": str | None}
  - None in from_zone_id matches ANY previous zone (including None/"no zone").
  - None in to_zone_id  matches ANY new zone (including None/"no zone").
  - None in asset_id    matches ANY asset.

TriggeredAlert (appended to alert_log when a rule fires):
  {"rule_id": str, "asset_id": str, "from_zone_id": str | None,
   "to_zone_id": str | None, "timestamp": str}

TrackerState:
  {
    "zones":       {zone_id: Zone},
    "assets":      {asset_id: Asset},
    "alert_rules": [AlertRule],
    "alert_log":   [TriggeredAlert],
  }
=============================================================================
"""

from typing import Optional


def make_tracker() -> dict:
    """Return a fresh, empty TrackerState."""
    return {"zones": {}, "assets": {}, "alert_rules": [], "alert_log": []}


# ---------------------------------------------------------------------------
# PART 1 — Zone membership  (warm-up, ~5 min)
# ---------------------------------------------------------------------------

def is_in_zone(asset: dict, zone: dict) -> bool:
    """
    Return True if the asset's lat/lng falls inside the zone's bounding box.
    - Bounds are inclusive on all edges.
    - Return False if the asset has no location (lat or lng is None).
    """
    raise NotImplementedError


# ---------------------------------------------------------------------------
# PART 2 — Locate an asset  (~5 min)
# ---------------------------------------------------------------------------

def get_current_zone_id(state: dict, asset_id: str) -> Optional[str]:
    """
    Return the id of the first zone in state that contains the asset, or None.
    - Return None if the asset doesn't exist or has no location.
    - Iterate zones in insertion order (standard Python dict behaviour).
    """
    raise NotImplementedError


# ---------------------------------------------------------------------------
# PART 3 — Process a location update  (core logic, ~15 min)
# ---------------------------------------------------------------------------

def process_location_update(
    state: dict,
    asset_id: str,
    lat: float,
    lng: float,
    timestamp: str,
) -> list:
    """
    Handle a new GPS reading for an asset:

    1. Update the asset's lat and lng in state.
    2. Recompute zone_id via get_current_zone_id and store it on the asset.
    3. If zone_id changed (including None→zone or zone→None), evaluate all
       alert_rules and collect any that match.
    4. Append each matched rule as a TriggeredAlert to state["alert_log"].
    5. Return the list of newly triggered alerts (empty list if no zone change
       or no rule matches).

    Raise KeyError if asset_id is not in state["assets"].

    Alert-rule matching — a rule matches when ALL three conditions hold:
        rule["asset_id"]     is None  OR  rule["asset_id"]     == asset_id
        rule["from_zone_id"] is None  OR  rule["from_zone_id"] == old_zone_id
        rule["to_zone_id"]   is None  OR  rule["to_zone_id"]   == new_zone_id
    """
    raise NotImplementedError


# ---------------------------------------------------------------------------
# PART 4 — CRUD helpers  (~15 min)
# ---------------------------------------------------------------------------

def add_zone(
    state: dict,
    zone_id: str,
    name: str,
    min_lat: float,
    max_lat: float,
    min_lng: float,
    max_lng: float,
) -> dict:
    """
    Create a zone, store it in state, and return it.
    Raise ValueError if zone_id already exists.
    """
    raise NotImplementedError


def remove_zone(state: dict, zone_id: str) -> None:
    """
    Remove a zone from state. Raise KeyError if not found.
    Any asset currently assigned to the removed zone should have its
    zone_id set to None. Do NOT fire alert rules for this forced change.
    """
    raise NotImplementedError


def add_asset(state: dict, asset_id: str, name: str) -> dict:
    """
    Create an asset with no initial location (lat=None, lng=None, zone_id=None),
    store it in state, and return it.
    Raise ValueError if asset_id already exists.
    """
    raise NotImplementedError


def add_alert_rule(
    state: dict,
    rule_id: str,
    from_zone_id: Optional[str],
    to_zone_id: Optional[str],
    asset_id: Optional[str],
) -> dict:
    """
    Add an alert rule to state and return it.
    Raise ValueError if rule_id already exists.
    """
    raise NotImplementedError

File: typescript/practice_problems/problem_01_geofence_alert_engine.ts
/**
 * =============================================================================
 * INTERVIEW PROBLEM 1: Geofence Alert Rule Engine
 * Difficulty: Senior Software Engineer | Estimated time: 45 min
 * =============================================================================
 *
 * CONTEXT
 * -------
 * You're building a backend service for an IoT asset-tracking platform. Physical
 * assets (forklifts, shipping containers, field equipment) carry GPS sensors that
 * periodically report coordinates. The platform tracks which geographic "zone"
 * (geofence) each asset is currently inside, and fires configured alert rules
 * whenever an asset transitions between zones.
 *
 * For this problem, zones are axis-aligned bounding boxes — no geospatial
 * libraries needed.
 *
 * DATA MODEL
 * ----------
 * All state lives in a single TrackerState object (returned by makeTracker).
 * Store all state in the TrackerState object itself — there is no class here,
 * every function takes state in and mutates/reads it explicitly.
 *
 * # Example
 * const state = makeTracker();
 * addZone(state, "warehouse", "Warehouse A", 35.0, 35.1, -106.7, -106.6);
 * addAsset(state, "forklift_1", "Forklift #1");
 * addAlertRule(state, "rule_entry", undefined, "warehouse", undefined);
 * processLocationUpdate(state, "forklift_1", 35.05, -106.65, "t1");
 * // -> [{ ruleId: "rule_entry", assetId: "forklift_1", fromZoneId: undefined,
 * //       toZoneId: "warehouse", timestamp: "t1" }]
 * =============================================================================
 */

export type Zone = {
  id: string;
  name: string;
  bounds: {
    minLat: number;
    maxLat: number;
    minLng: number;
    maxLng: number;
  };
};

export type Asset = {
  id: string;
  name: string;
  lat?: number;
  lng?: number;
  zoneId?: string;
};

export type AlertRule = {
  id: string;
  fromZoneId?: string;
  toZoneId?: string;
  assetId?: string;
};

export type TriggeredAlert = {
  ruleId: string;
  assetId: string;
  fromZoneId?: string;
  toZoneId?: string;
  timestamp: string;
};

export type TrackerState = {
  zones: Record<string, Zone>;
  assets: Record<string, Asset>;
  alertRules: AlertRule[];
  alertLog: TriggeredAlert[];
};

/** Return a fresh, empty TrackerState. */
export function makeTracker(): TrackerState {
  return { zones: {}, assets: {}, alertRules: [], alertLog: [] };
}

// ---------------------------------------------------------------------------
// PART 1 — Zone membership  (warm-up, ~5 min)
// ---------------------------------------------------------------------------

/**
 * Return true if the asset's lat/lng falls inside the zone's bounding box.
 * - Bounds are inclusive on all edges.
 * - Return false if the asset has no location (lat or lng is undefined).
 */
export function isInZone(asset: Asset, zone: Zone): boolean {
  throw new Error("Not implemented");
}

// ---------------------------------------------------------------------------
// PART 2 — Locate an asset  (~5 min)
// ---------------------------------------------------------------------------

/**
 * Return the id of the first zone in state that contains the asset, or undefined.
 * - Return undefined if the asset doesn't exist or has no location.
 * - Iterate zones in insertion order (standard object key order).
 */
export function getCurrentZoneId(state: TrackerState, assetId: string): string | undefined {
  throw new Error("Not implemented");
}

// ---------------------------------------------------------------------------
// PART 3 — Process a location update  (core logic, ~15 min)
// ---------------------------------------------------------------------------

/**
 * Handle a new GPS reading for an asset:
 *
 * 1. Update the asset's lat and lng in state.
 * 2. Recompute zoneId via getCurrentZoneId and store it on the asset.
 * 3. If zoneId changed (including undefined→zone or zone→undefined), evaluate
 *    all alertRules and collect any that match.
 * 4. Append each matched rule as a TriggeredAlert to state.alertLog.
 * 5. Return the list of newly triggered alerts (empty array if no zone change
 *    or no rule matches).
 *
 * Throw an Error if assetId is not in state.assets.
 *
 * Alert-rule matching — a rule matches when ALL three conditions hold:
 *     rule.assetId     is undefined  OR  rule.assetId     === assetId
 *     rule.fromZoneId  is undefined  OR  rule.fromZoneId  === oldZoneId
 *     rule.toZoneId    is undefined  OR  rule.toZoneId    === newZoneId
 */
export function processLocationUpdate(state: TrackerState, assetId: string, lat: number, lng: number, timestamp: string): TriggeredAlert[] {
  throw new Error("Not implemented");
}

// ---------------------------------------------------------------------------
// PART 4 — CRUD helpers  (~15 min)
// ---------------------------------------------------------------------------

/**
 * Create a zone, store it in state, and return it.
 * Throw an Error if zoneId already exists.
 */
export function addZone(state: TrackerState, zoneId: string, name: string, minLat: number, maxLat: number, minLng: number, maxLng: number): Zone {
  throw new Error("Not implemented");
}

/**
 * Remove a zone from state. Throw an Error if not found.
 * Any asset currently assigned to the removed zone should have its
 * zoneId set to undefined. Do NOT fire alert rules for this forced change.
 */
export function removeZone(state: TrackerState, zoneId: string): void {
  throw new Error("Not implemented");
}

/**
 * Create an asset with no initial location (lat=undefined, lng=undefined,
 * zoneId=undefined), store it in state, and return it.
 * Throw an Error if assetId already exists.
 */
export function addAsset(state: TrackerState, assetId: string, name: string): Asset {
  throw new Error("Not implemented");
}

/**
 * Add an alert rule to state and return it.
 * Throw an Error if ruleId already exists.
 */
export function addAlertRule(
  state: TrackerState,
  ruleId: string,
  fromZoneId: string | undefined,
  toZoneId: string | undefined,
  assetId: string | undefined,
): AlertRule {
  throw new Error("Not implemented");
}
